Kernel/Storage/NVMe/NVMeQueue.cpp

```diff
@@ -113,38 +105,40 @@ void NVMeQueue::submit_sqe(NVMeSubmission& sub)
 u16 NVMeQueue::submit_sync_sqe(NVMeSubmission& sub)
 {
     // For now let's use sq tail as a unique command id.
-    u16 cqe_cid;
-    u16 cid = m_sq_tail;
+    u16 cmd_status;
+    u16 cid = get_request_cid();
+    sub.cmdid = cid;
 
+    {
+        SpinlockLocker req_lock(m_request_lock);
+
+        if (m_requests.contains(sub.cmdid) && m_requests.get(sub.cmdid).release_value().used)
+            VERIFY_NOT_REACHED();
+        m_requests.set(sub.cmdid, { nullptr, true, [this, &cmd_status](u16 status) mutable { cmd_status = status; m_sync_wait_queue.wake_all(); } });
+    }
```

Comment:
it looks like this part could be abstracted (parametrized by a request and a callback) as it seems to be used in three places already… 🤔